logstash-core/src/test/java/org/logstash/config/ir/CompiledPipelineTest.java

```diff
@@ -159,12 +164,57 @@ public void buildsForkedPipeline() throws Exception {
                 filters,
                 Collections.singletonMap("mockoutput", mockOutputSupplier())
             )
-        ).buildExecution().compute(RubyUtil.RUBY.newArray(testEvent), false, false);
+        ).buildExecution(false).compute(RubyUtil.RUBY.newArray(testEvent), false, false);
         final Collection<JrubyEventExtLibrary.RubyEvent> outputEvents = EVENT_SINKS.get(runId);
         MatcherAssert.assertThat(outputEvents.size(), CoreMatchers.is(1));
         MatcherAssert.assertThat(outputEvents.contains(testEvent), CoreMatchers.is(true));
     }
 
+    @Test
+    public void preservesInputEventInstantiationOrder() throws Exception {
+        testEventOrder(true, new int[]{0, 2, 1});
+    }
+
+    @Test
+    public void ignoresInputEventInstantiationOrder() throws Exception {
+        testEventOrder(false, null);
+    }
+
+    @SuppressWarnings("rawtypes")
+    private void testEventOrder(boolean sortedEvents, int[] expectedOrdering) throws Exception {
+        final RubyEvent event1 = RubyEvent.newRubyEvent(RubyUtil.RUBY, new Event());
+        event1.getEvent().setField("message", 1);
+        final RubyEvent event2 = RubyEvent.newRubyEvent(RubyUtil.RUBY, new Event());
+        event2.getEvent().setField("message", 2);
+        event2.getEvent().tag("foo");
+        final RubyEvent event3 = RubyEvent.newRubyEvent(RubyUtil.RUBY, new Event());
+        event3.getEvent().setField("message", 3);
+
+        // arrange them out-of-order in the input batch to test ordering
+        RubyArray inputBatch = RubyUtil.RUBY.newArray(event1, event3, event2);
```

Comment:
I'm not sure that manually constructing an out-of-order batch verifies the behaviour we are attempting to guarantee, which is that events that may be reordered by the result of an `if-branch` are zipped back together in order before being handled by the next node.

Reply:
I updated the test so that a filter node with a conditional causes the reordering of events.